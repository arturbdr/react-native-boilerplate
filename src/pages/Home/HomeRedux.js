const GET_HOME = 'GET_HOME';
const SET_HOME = 'SET_HOME';

/* Actions ------------- */
const get = () => ({ type: GET_HOME });
const success = (home: Object) => ({ type: SET_HOME, home });

type Action = $Call<typeof get> | $Call<typeof success, Object>;

/* Reducer ------------------- */
const initialState = {
  home: {},
};

const reducer = (state: typeof initialState = initialState, action: Action) => {
  switch (action.type) {
    case GET_HOME:
      return { ...state };
    case SET_HOME:
      return { ...state, home: { ...action.home } };
    default:
      return state;
  }
};

export { get, success, reducer, GET_HOME, SET_HOME };
