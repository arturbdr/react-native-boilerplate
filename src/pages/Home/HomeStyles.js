import styled from 'styled-components';

const Container = styled.View`
  flex: 1;
  background-color: ${props => props.theme.background.primary};
`;

const Button = styled.Button`
  margin-bottom: 10;
`;

const Text = styled.Text`
  color: ${props => props.theme.text.primary};
`;

export { Button, Container, Text };
