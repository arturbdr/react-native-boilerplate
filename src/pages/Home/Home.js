import React from 'react';
import type { NavigationScreenProp, NavigationState } from 'react-navigation';
import { bindActionCreators as bind, type Dispatch } from 'redux';
import { translate, type TFunction } from 'react-i18next';
import { compose } from 'recompose';
import { connect } from 'react-redux';

import { Container, Button, Text } from './HomeStyles';
import withHeader from '../../commons/hocs/withHeader';
import * as Actions from './HomeRedux';

export type HomeType = {
  category: string,
  icon_url: string,
  id: string,
  url: string,
  value: string,
};

type PropsType = {
  t: TFunction,
  i18n: Object,
  navigation: NavigationScreenProp<NavigationState>,
  home: HomeType,
  get: typeof Actions.get,
};

export const HomeBase = ({ t, i18n, navigation, home, get }: PropsType) => (
  <Container>
    <Text>{t('common:currentLanguage', { lng: 'language' })}</Text>
    <Text>{t('home:title')}</Text>
    <Text>{t('home:introduction')}</Text>
    <Button
      block
      title={t('common:actions.toggleToEnglish')}
      onPress={() => {
        i18n.changeLanguage('en');
      }}
    />
    <Button
      block
      title={t('common:actions.toggleToPortuguese')}
      onPress={() => {
        i18n.changeLanguage('pt');
      }}
    />
    <Button
      block
      title={t('common:actions.goToPage2')}
      onPress={() => navigation.navigate('NextPage')}
    />
    <Button block onPress={get} title={t('common:actions.fetchData')} />
    <Text>{JSON.stringify(home)}</Text>
  </Container>
);

export default compose(
  translate(['home', 'common']),
  connect(
    state => ({ ...state.home }),
    (dispatch: Dispatch<*>) => bind({ get: Actions.get }, dispatch)
  ),
  withHeader(({ t }: PropsType) => ({
    title: t('home:headerTitle'),
  }))
)(HomeBase);
