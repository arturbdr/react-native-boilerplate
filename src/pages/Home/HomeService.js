import { type HomeType } from './Home';

type HomeApiType = {
  category?: string,
  icon_url: string,
  id: string,
  url: string,
  value: string,
};

class HomeService {
  static mapApiToValues(home: HomeApiType): HomeType {
    return {
      ...home,
      category: home.category || '',
    };
  }

  // @FIXME implement or remove
  static mapValuesToApi() {}
}

export default HomeService;
