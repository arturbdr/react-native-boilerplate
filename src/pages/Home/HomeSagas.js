import { type Saga } from 'redux-saga';

import { takeEvery, call, put } from 'redux-saga/effects';
import { modules, routes } from '../../config/constants';
import { get } from '../../config/service';
import { success, GET_HOME } from './HomeRedux';
import HomeService from './HomeService';

function* getSaga(): Saga<void> {
  try {
    const result = yield call(get, modules.jokes, routes.random);
    const resultParsed = HomeService.mapApiToValues(result);
    yield put(success(resultParsed));
  } catch (error) {
    // Toast.show({ text: error.userMessage }); @FIXME show error
  }
}

export default [takeEvery(GET_HOME, getSaga)];
