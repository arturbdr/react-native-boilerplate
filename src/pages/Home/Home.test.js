import React from 'react';

import { HomeBase } from './Home';

function setup(injectedProps = {}) {
  const props = {
    t: jest.fn(),
    navigation: jest.fn(),
    ...injectedProps,
  };
  const wrapper = shallow(<HomeBase {...props} />, {
    disableLifecycleMethods: false,
  });

  return {
    props,
    wrapper,
  };
}

describe('Visit', () => {
  it('renders without crashing', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
