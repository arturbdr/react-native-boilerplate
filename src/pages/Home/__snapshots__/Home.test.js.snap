// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visit renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HomeBase
    navigation={[MockFunction]}
    t={
      [MockFunction] {
        "calls": Array [
          Array [
            "common:currentLanguage",
            Object {
              "lng": "language",
            },
          ],
          Array [
            "home:title",
          ],
          Array [
            "home:introduction",
          ],
          Array [
            "common:actions.toggleToEnglish",
          ],
          Array [
            "common:actions.toggleToPortuguese",
          ],
          Array [
            "common:actions.goToPage2",
          ],
          Array [
            "common:actions.fetchData",
          ],
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Styled(Text) />,
        <Styled(Text) />,
        <Styled(Text) />,
        <Styled(Button)
          block={true}
          onPress={[Function]}
          title={undefined}
        />,
        <Styled(Button)
          block={true}
          onPress={[Function]}
          title={undefined}
        />,
        <Styled(Button)
          block={true}
          onPress={[Function]}
          title={undefined}
        />,
        <Styled(Button)
          block={true}
          onPress={undefined}
          title={undefined}
        />,
        <Styled(Text) />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "block": true,
          "onPress": [Function],
          "title": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "block": true,
          "onPress": [Function],
          "title": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "block": true,
          "onPress": [Function],
          "title": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "block": true,
          "onPress": undefined,
          "title": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Styled(Text) />,
          <Styled(Text) />,
          <Styled(Text) />,
          <Styled(Button)
            block={true}
            onPress={[Function]}
            title={undefined}
          />,
          <Styled(Button)
            block={true}
            onPress={[Function]}
            title={undefined}
          />,
          <Styled(Button)
            block={true}
            onPress={[Function]}
            title={undefined}
          />,
          <Styled(Button)
            block={true}
            onPress={undefined}
            title={undefined}
          />,
          <Styled(Text) />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "block": true,
            "onPress": [Function],
            "title": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "block": true,
            "onPress": [Function],
            "title": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "block": true,
            "onPress": [Function],
            "title": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "block": true,
            "onPress": undefined,
            "title": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": false,
    "hydrateIn": undefined,
  },
}
`;
